go/createbucket.go

```diff
@@ -0,0 +1,103 @@
+package bucketclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+type createBucketOptionSet struct {
+	sessionId      int
+	makeIdempotent bool
+}
+
+type CreateBucketOption func(*createBucketOptionSet)
+
+func CreateBucketSessionIdOption(sessionId int) CreateBucketOption {
+	return func(options *createBucketOptionSet) {
+		options.sessionId = sessionId
+	}
+}
+
+func CreateBucketMakeIdempotent(options *createBucketOptionSet) {
+	options.makeIdempotent = true
+}
+
+// CreateBucket creates a bucket in metadata.
+// bucketAttributes is a JSON blob of bucket attributes
+// opts is a set of options:
+//
+//	CreateBucketSessionIdOption forces the session ID where the bucket to be
+//	    created will land
+//
+//	CreateBucketMakeIdempotent makes the request return a success if a bucket
+//	    with the same UID already exists (otherwise returns 409 Conflict, as
+//	    if the option is not passed)
+func (client *BucketClient) CreateBucket(ctx context.Context,
+	bucketName string, bucketAttributes []byte, opts ...CreateBucketOption) error {
```

Comment:
See comment below, but I'm not too keen on carrying `[]byte` around when we should be defining proper struct instead. I know it's more cumbersome but it's a real investment in the lib and will make it much more stable and easy to use.

Reply:
At this level, I think the bucket attributes should be manipulated as an opaque binary blob, in theory, although I strayed a bit to be able to compare the UIDs and avoid overwriting the wrong bucket, but I'd prefer to limit the parsing to just this.

The bucket attributes is really a cloudserver-specific object, there are other metadata in the map that are known to bucketd to fit its purpose (i.e know the raft session ID etc.). Ideally a UID would also be added there for bucketd to check, that could be a different one than the UID in the attributes.